odl/discr/partition.py

```diff
@@ -546,6 +555,139 @@ def squeeze(self):
+        >>> p = uniform_partition(0, 2, 5)
+        >>> p.index(0)
+        0
+        >>> p.index(2)
+        4
+
+        Also supports points inside voxels, in that case returning the index
+        of the voxel:
+
+        >>> p.index(0.2)
+        0
+
+        By using the ``floating`` argument, partial positions inside the voxels
+        can instead be determined:
+
+        >>> p.index(0.2, floating=True)
+        0.5
+
+        These indexes work with indexing, extracting the voxel the point lies
+        in:
+
+        >>> p[p.index(0.1)]
+        uniform_partition(0.0, 0.4, 1)
+
+        Index also works in higher dimensions:
+
+        >>> p = uniform_partition([0, -1], [1, 2], (4, 1))
+        >>> p.index([0.5, 2])
+        (2, 0)
+        >>> p[p.index([0.5, 2])]
+        uniform_partition([0.5, -1.0], [0.75, 2.0], [1, 1])
+        """
+        value = np.atleast_1d(self.set.element(value))
+        result = ()
+        for val, cell_bdry_vec in zip(value, self.cell_boundary_vecs):
+            ind = np.searchsorted(cell_bdry_vec, val)
+            if floating:
+                if cell_bdry_vec[ind] == val:
+                    # Value is on top of edge
+                    result += (float(ind),)
+                else:
+                    # interpolate between
+                    csize = float(cell_bdry_vec[ind] - cell_bdry_vec[ind - 1])
+                    result += (ind - (cell_bdry_vec[ind] - val) / csize,)
+            else:
+                if cell_bdry_vec[ind] == val and ind != len(cell_bdry_vec) - 1:
+                    # Value is on top of edge, but not last edge
+                    result += (ind,)
+                else:
+                    result += (ind - 1,)
+
+        if self.ndim == 1:
+            result = result[0]
+
+        return result
+
+    @property
+    def bydimension(self):
+        """Return the subpartition defined along one or several dimensions.
+
+        Examples
+        --------
+        Can get each dimension independently:
+
+        >>> p = uniform_partition([0, 1, 2], [1, 3, 5], [3, 5, 6])
+        >>> p.bydimension[0]
+        uniform_partition(0.0, 1.0, 3)
+        >>> p.bydimension[1]
+        uniform_partition(1.0, 3.0, 5)
+        >>> p.bydimension[2]
+        uniform_partition(2.0, 5.0, 6)
+
+        Usual numpy style advanced indexing works:
+
+        >>> p.bydimension[:] == p
+        True
+        >>> p.bydimension[1:]
+        uniform_partition([1.0, 2.0], [3.0, 5.0], [5, 6])
+        >>> p.bydimension[[0, 2]]
+        uniform_partition([0.0, 2.0], [1.0, 5.0], [3, 6])
+        """
+        partition = self
+
+        class RectPartitionByDimension(object):
+            """Helper class for accessing `RectPartition` by dimension.
+
+            See Also
+            --------
+            RectPartition.bydimension
+            """
+
+            def __getitem__(self, dim):
+                """Return ``self[dim]``."""
+                slc = np.zeros(partition.ndim, dtype=object)
```

Comment:
> but what about the dtype=object thing?

Without it I cannot do

```
slc[dim] = np.s_[:]
```

later on. If the dtype was `int`, then assigning a `slice` would fail.